### DISCUSSION TITLE
metric-tensor's general questions

### DISCUSSION THREAD

**Discussion by metric-tensor** [RavEngine/RavEngine#5] [Q&A]
I am playing around with RavEngine for a hobby space sim project.  I just wanted to start a thread for some ongoing questions

1. What is the difference between  a world and a scene?
2. I am happy to see support for double precision in mathtypes.hpp but am also aware that PhysX doesn't support double precision, how is this handled?  Have you considered a physics engine with double precision support?
3. What is the difference between a static mesh and a mesh asset?
4. Are their built-in mesh generators?  ( icosphere, etc )

**Comment by Ravbug** [comment]
Hey, thanks for the questions!

1. What Unity calls a Scene is what RavEngine calls a World. Worlds contain all the action in a game. 
2. In essence it makes the transform component and the different linear algebra types use double instead of float. Physics and rendering still operate in single-precision mode. I haven't tested the double-precision mode in quite a while though, so you might get odd or broken behavior when using it. Switching to a different physics engine is currently not on the roadmap.
3. MeshAsset refers to the vertex / index buffers representing the geometry data, while StaticMesh is a component that you attach to your entities, so that it renders in the world with a transform and a material. In general, Assets are not directly connected to entities. 
4. There are a few hardcoded mesh primitives (cube, cone, triangle sphere), but no generators atm. To add this feature within your application without needing to modify engine code, it probably wouldn't be very difficult to write or import a 3rd party library that can generate meshes and use the vertex/index constructors of MeshAsset.

I have a few code samples which you might find helpful:
- [RavEngine/HelloCube](https://github.com/RavEngine/HelloCube) - a minimal sample which illustrates via code comments how to spawn and animate a cube 
- [RavEngine/Samples](https://github.com/RavEngine/Samples) - the bulk of the code samples, which double as my test cases and highlight other features of the engine like physics, audio, and networking.

Hope this helps!

**Comment by metric-tensor** [comment]
I am not familiar with Unity ( and its scenes), I just see that there is a `SceneLoader` class.  Does this imply that the `SceneLoader` loads world objects?

I have been looking at some of the examples, I will dig deeper.

**Comment by Ravbug** [comment]
The `SceneLoader` is for importing files like .fbx. I agree that the naming is a little confusing. At the moment, worlds are created solely by constructing them and calling their `CreatePrototype<T>` method (which will soon be renamed to Instantiate) to populate them. It's often convenient to subclass `RavEngine::World` to do this, which is why the samples do it. 

You can see SceneLoader usage in the [SceneViewer example](https://github.com/RavEngine/Samples/blob/master/Samples/SceneViewer/).

**Comment by metric-tensor** [comment]
What's the impetus for going away from BGFX?

**Comment by Ravbug** [comment]
I want to provide a better experience for writing and debugging shaders than is possible with bgfx. I also want to support things like hardware raytracing in the future, which bgfx does not support.

**Comment by metric-tensor** [comment]
When I run the _RavEnginePerf_DrawAnimated_ sample, the TPS is only 4 or 5.  The other samples seem to run at a reasonable rate.  Is this normal behavior?  (Win 11)

**Reply by Ravbug** [reply]
It is expected for Perf-DrawAnimated to run poorly, because it is a stress test designed to help me locate inefficiencies in the engine.

What hardware are you running on?

**Reply by metric-tensor** [reply]
Ryzen 9 7950X, 128GB RAM
RTX 3080

**Comment by metric-tensor** [comment]
Does RavEngine have a particle system?

**Reply by Ravbug** [reply]
Not currently, but I am working on plans for a GPU particle system. No ETA though

**Comment by PeanutButterGaming** [comment]
is there a tutorial on how to set up the engine for Visual Studio? I'm new to all of this and learning c++ atm.
edit: i know see the litteral tutorial for building it on the main code page. nvm

**Reply by PeanutButterGaming** [reply]
also, is there any plans for a visual, like, ui interface in the future?

**Reply by RallyTronics** [reply]
Make sure you remember this part when you build "On Windows, you will need to run your initial configure twice before building (known bug)."

**Reply by Ravbug** [reply]
> also, is there any plans for a visual, like, ui interface in the future?

There are currently no plans for a visual editor

**Reply by PeanutButterGaming** [reply]
Is there a more in-depth tutorial for installation? I have no idea what cmake is, nor 'building'. so what do i do?

**Reply by Ravbug** [reply]
> I have no idea what cmake is, nor 'building'. so what do i do?

In that case, this is probably not the engine for you. The intended audience for this engine is experienced C++ developers. 

I would highly recommend checking out Godot if you are new to development and want an open-source beginner-friendly game engine.

**Reply by PeanutButterGaming** [reply]
:|

**Reply by PeanutButterGaming** [reply]
I can learn.

**Comment by brightening-eyes** [comment]
hello,
thanks for such an awesome and small engine.
I have some questions:
* How would I create a character controller, vehicle etc on ravengine?
* also, like [ezengine](https://github.com/ezengine/ezengine), is there any way possible that I can write a component (here an entity I suppose) that can be updated every frame? for example, a vehicle that can be moved around/or controlled by user (maybe the player's component is attached to it then player needs to have some ways of control etc).
* how to have continuous collision detection (ccd)? this can be used for shooting bullets/projectiles, checking if something is in range etc.
* what is the coordinate system of RavEngine?

thanks

**Reply by Ravbug** [reply]
Hey, thanks for checking out this project! 
1. I have a character controller example in the Samples repo: https://github.com/RavEngine/Samples/blob/master/Samples/Animation/Character.cpp
2. RavEngine uses an ECS. First, you'd create an entity, attach components to the entity, and then provide a System which filters the world and can update values in components. You can see a simple example of this in the Samples repo: https://github.com/RavEngine/Samples/blob/master/Samples/Perf_ECS/Level.cpp
3. To get CCD, you'd need to make some source changes to the engine to enable that feature in PhysX, since I currently don't expose it. That shouldn't be too hard to do though. https://docs.nvidia.com/gameworks/content/gameworkslibrary/physx/guide/Manual/AdvancedCollisionDetection.html
4. +Y-up, -Z forward, and +X-right. The depth texture uses inverse-Z.

**Comment by brightening-eyes** [comment]
hello again,
thanks for the answers.
* for something like updating physics world etc, I saw that ctti is being used.
my question is, how is it supposed to be called? for physicsLinkSystem and ScriptSystem it works differently. now, if someone wants to make another system, lets say vegetation system and/or mesh generation plugin or a sensor component which can track things or check weather something is in its range etc.
* in the character example, a rigid body was used I suppose. now, physX has character controller which can be integrated. the same is true with vehicles. I mean how these are supposed to be implemented into my code? where do I need to make source changes to the engine if required (and I will make a pull request for it as well)
* another thing is different shapes.
how would assign different shapes to physics components?
I mean how can I make compound shapes and then assign that into my body?
* what about sensor shapes? (no mesh component, but just a sensor component that can check if something is in range). like an electronic door that checks weather someone is in its range and opens it, otherwise closes it.
thanks.

**Reply by Ravbug** [reply]
- I highly recommend reading about how Entity-Component-Systems work, since this engine uses one at its core, and all gameplay functionality is implemented through it. Then, the examples in the Samples repo will probably make more sense. The [HelloCube](https://github.com/RavEngine/HelloCube) minimal example shows how to create a custom System and update objects using it. 
- To properly add the PhysX character controller or PhysX vehicles as an engine feature (i.e. as something I would accept as a PR), you'd need to make a `PhysicsCharacterComponent` or `PhysicsVehicleComponent` and add any logic needed to interface it with the existing `PhysicsSolver` class. You'd also need to make a new example for the Samples repo showing it working as that will serve as both a test case and as documentation. 
- The [AirHockey](https://github.com/RavEngine/Samples/tree/master/Samples/AirHockey) example shows how to add collision shapes to physics components, including adding multiple shapes to the same body.
- To make a Sensor, you'd probably want to create a static rigid body with a trigger collider. Alternatively, you could forgo the physics system and implement a custom System which uses the `Filter` feature of Worlds to find objects, get their Transforms, and check distances to the current object.
